common/throttle.h

```diff
@@ -33,23 +34,61 @@ class throttle {
      * @param limit -1UL means no limit, 0 means lowest speed (hang)
      */
     explicit throttle(uint64_t limit, uint64_t time_window = 1000UL * 1000,
-             uint64_t slice = 10) : m_slice_num(slice) {
+             uint64_t slice = 100) : m_slice_num(slice) {
         update(limit);
+        // Equals to DIV_ROUND_UP
         m_time_window_per_slice = photon::sat_add(time_window, (m_slice_num - 1)) / m_slice_num;
         m_time_window = m_time_window_per_slice * m_slice_num;
+        for (auto& each: m_starving_slice_num) each = 0;
+        int i = 0;
+        for (auto& each: m_starving_slice_percent) each = get_starving_percent(Priority(i++));
         sem.signal(m_limit);
     }
 
+    enum class Priority {
+        High,
+        Medium,
+        Low,
+        NumPriorities,
+    };
+
     void update(uint64_t limit) {
+        // Equals to DIV_ROUND_UP
         m_limit_per_slice = photon::sat_add(limit, (m_slice_num - 1)) / m_slice_num;
         m_limit = m_limit_per_slice * m_slice_num;
     }
 
-    int consume(uint64_t amount) {
-        int ret = 0;
-        int err = 0;
+    int consume(uint64_t amount, Priority prio = Priority::High) {
```

Comment:
Should `restore` and `try_consume` works with priority?

One of the common throttle usage used to be
```
...
throttle_control.try_consume(amount);
ret = do_some_work;
if (ret.failed()) {
  restore(amount);
}
```

To consume resource before actually present action, can limit boost concurrent actions push through limits.
and to restore unconsumed resources if failed or partial succeed, so failed action may not count in throughput limits.

Reply:
I have thought about it again, `try_consume` may not need priority.